File: src/main/java/Piruru/interfaces/OnRefreshHand.java
package Piruru.interfaces;

import Piruru.Piruru;
import Piruru.stances.Allos;
import com.evacipated.cardcrawl.modthespire.lib.*;
import com.megacrit.cardcrawl.cards.CardGroup;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.powers.AbstractPower;
import javassist.CtBehavior;

public interface OnRefreshHand {

    public void onRefreshHand();


    @SpirePatch(
            clz = CardGroup.class,
            method = "refreshHandLayout"
    )
    public class OnRefreshHandPatch {
        @SpireInsertPatch(
                locator = Locator.class
        )
        public static void Insert(CardGroup __instance) {
            for (AbstractPower p : AbstractDungeon.player.powers) {
                if (p instanceof OnRefreshHand) {
                    ((OnRefreshHand) p).onRefreshHand();
                }
            }
            if (AbstractDungeon.player.stance instanceof Allos) {
                ((OnRefreshHand) AbstractDungeon.player.stance).onRefreshHand();
            }
        }

        public static class Locator extends SpireInsertLocator {

            @Override
            public int[] Locate(CtBehavior ctMethodToPatch) throws Exception {
                Matcher finalMatcher = new Matcher.FieldAccessMatcher(CardGroup.class, "group");
                return LineFinder.findInOrder(ctMethodToPatch, finalMatcher);
            }
        }
    }
    
}

File: src/main/java/Piruru/interfaces/OnRefreshHand.kt
package Piruru.interfaces

import Piruru.stances.Allos
import com.evacipated.cardcrawl.modthespire.lib.*
import com.megacrit.cardcrawl.cards.CardGroup
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.powers.AbstractPower
import javassist.CtBehavior

interface OnRefreshHand {

    fun onRefreshHand()


    @SpirePatch(clz = CardGroup::class, method = "refreshHandLayout")
    object OnRefreshHandPatch {
        @SpireInsertPatch(locator = Locator::class)
        fun Insert(__instance: CardGroup) {
            for (p in AbstractDungeon.player.powers) {
                if (p is OnRefreshHand) {
                    (p as OnRefreshHand).onRefreshHand()
                }
            }
            if (AbstractDungeon.player.stance is Allos) {
                (AbstractDungeon.player.stance as OnRefreshHand).onRefreshHand()
            }
        }

        class Locator : SpireInsertLocator() {

            @Throws(Exception::class)
            override fun Locate(ctMethodToPatch: CtBehavior): IntArray {
                val finalMatcher = Matcher.FieldAccessMatcher(CardGroup::class.java, "group")
                return LineFinder.findInOrder(ctMethodToPatch, finalMatcher)
            }
        }
    }

}
